packages/core/src/common/message-rpc/rpc-protocol.ts

```diff
@@ -132,15 +132,11 @@ export class RpcProtocol {
         // The last element of the request args might be a cancellation token. As these tokens are not serializable we have to remove it from the
         // args array and the `CANCELLATION_TOKEN_KEY` string instead.
         const cancellationToken: CancellationToken | undefined = args.length && CancellationToken.is(args[args.length - 1]) ? args.pop() : undefined;
-        if (cancellationToken && cancellationToken.isCancellationRequested) {
-            return Promise.reject(this.cancelError());
-        }
```

Comment:
From the comment above about subbing in a magic string for the cancellation token itself, it looks like we don't pass the state of token - whether it's already been cancelled or not - to the receiving end. If that's the case, then by removing this block, we remove handling the case in which the token has already been cancelled when we get here, since I don't believe that it will fire the `onCancellationRequested` event again. In that case, should we send the cancellation notice immediately after sending the request itself?

Reply:
Good point. Yes, sending the cancellation notice directly after sending the request sounds reasonable to me.

Reply:
@colin-grant-work After investigating and testing this it turns out that both main `CancellationToken` implementations (Theia and vscode-languageserver) already have a special shortcut event in place that will trigger newly registered cancellation listeners on already canceled tokens immediately. So the cancellation notice for immediately canceled tokens is already passed correctly to the receiving side.

However, just to be on the safe side (and since in theory also custom cancellation token implementations could be used)    I still added the discussed behavior. Directly after sending the request we check whether the token is already canceled.
If so, send the cancellation notice, otherwise listen on the `onCancellationRequested` event.